crates/assists/src/handlers/qualify_path.rs

```diff
@@ -0,0 +1,1057 @@
+        }
+        ImportCandidate::TraitAssocItem(_) => {
+            let path = ast::Path::cast(import_assets.syntax_under_caret().clone())?;
+            let (qualifier, segment) = (path.qualifier()?, path.segment()?);
+            qualify_path_trait_assoc_item(acc, proposed_imports, range, qualifier, segment)
+        }
+        ImportCandidate::TraitMethod(_) => {
+            let mcall_expr = ast::MethodCallExpr::cast(import_assets.syntax_under_caret().clone())?;
+            qualify_path_trait_method(acc, ctx.sema.db, proposed_imports, range, mcall_expr)?;
+        }
+    };
+    Some(())
+}
+
+// a test that covers this -> `associated_struct_const`
+fn qualify_path_qualifier_start(
+    acc: &mut Assists,
+    proposed_imports: Vec<(hir::ModPath, hir::ItemInNs)>,
+    range: TextRange,
+    segment: ast::PathSegment,
+    qualifier_start: &ast::NameRef,
+) {
+    mark::hit!(qualify_path_qualifier_start);
+    let group_label = GroupLabel(format!("Qualify {}", qualifier_start));
+    for (import, _) in proposed_imports {
+        acc.add_group(
+            &group_label,
+            AssistId("qualify_path", AssistKind::QuickFix),
+            format!("Qualify with `{}`", &import),
+            range,
+            |builder| {
+                let import = mod_path_to_ast(&import);
+                builder.replace(range, format!("{}::{}", import, segment));
+            },
+        );
+    }
+}
+
+// a test that covers this -> `applicable_when_found_an_import_partial`
+fn qualify_path_unqualified_name(
+    acc: &mut Assists,
+    proposed_imports: Vec<(hir::ModPath, hir::ItemInNs)>,
+    range: TextRange,
+    name: &ast::NameRef,
+) {
+    mark::hit!(qualify_path_unqualified_name);
+    let group_label = GroupLabel(format!("Qualify {}", name));
+    for (import, _) in proposed_imports {
+        acc.add_group(
+            &group_label,
+            AssistId("qualify_path", AssistKind::QuickFix),
+            format!("Qualify as `{}`", &import),
+            range,
+            |builder| builder.replace(range, mod_path_to_ast(&import).to_string()),
+        );
+    }
+}
+
+// a test that covers this -> `associated_trait_const`
+fn qualify_path_trait_assoc_item(
+    acc: &mut Assists,
+    proposed_imports: Vec<(hir::ModPath, hir::ItemInNs)>,
+    range: TextRange,
+    qualifier: ast::Path,
+    segment: ast::PathSegment,
+) {
+    mark::hit!(qualify_path_trait_assoc_item);
+    let group_label = GroupLabel(format!("Qualify {}", &segment));
+    for (import, _) in proposed_imports {
+        acc.add_group(
+            &group_label,
+            AssistId("qualify_path", AssistKind::QuickFix),
+            format!("Qualify with cast as `{}`", &import),
+            range,
+            |builder| {
+                let import = mod_path_to_ast(&import);
+                builder.replace(range, format!("<{} as {}>::{}", qualifier, import, segment));
+            },
+        );
```

Comment:
Hm, can we refactor this so that there's only one call to `add_group`? We currently repeat `AssistId` three times and that feels less then ideal. Though, this doesn't matter much, if there isn't easy way out here, current code is fine.

Reply:
The `AssistId` could be made a const.

Reply:
I mean, yes, that would brute-force fix the duplication. But I suspect that the overall shape of code would be better if it reads

```
add_assist(
  if cond1 {}
  if cond2 {}
  if cond2 {}
)
```

rather than 

```
if cond1 { add_assist() }
if cond2 { add_assist() }
if cond3 { add_assist() }
```

The goal is not to reduce textual duplication, but rather simplify the overall shape of the code.

Reply:
Ah okay that is what you meant, my bad 😄. I'll take a look at that then.

Reply:
Alright I made some changes to de-duplicate the add_group calls